src/gt4py/next/ffront/foast_to_gtir.py

```diff
@@ -0,0 +1,384 @@
+        elif isinstance(node.func, foast.Name) and node.func.id in fbuiltins.MATH_BUILTIN_NAMES:
+            return self._visit_math_built_in(node, **kwargs)
+        elif isinstance(node.func, foast.Name) and node.func.id in (
+            fbuiltins.FUN_BUILTIN_NAMES + experimental_builtins.EXPERIMENTAL_FUN_BUILTIN_NAMES
+        ):
+            visitor = getattr(self, f"_visit_{node.func.id}")
+            return visitor(node, **kwargs)
+        elif isinstance(node.func, foast.Name) and node.func.id in fbuiltins.TYPE_BUILTIN_NAMES:
+            return self._visit_type_constr(node, **kwargs)
+        elif isinstance(
+            node.func.type,
+            (ts.FunctionType, ts_ffront.FieldOperatorType, ts_ffront.ScanOperatorType),
+        ):
+            # ITIR has no support for keyword arguments. Instead, we concatenate both positional
+            # and keyword arguments and use the unique order as given in the function signature.
+            lowered_args, lowered_kwargs = type_info.canonicalize_arguments(
+                node.func.type,
+                self.visit(node.args, **kwargs),
+                self.visit(node.kwargs, **kwargs),
+                use_signature_ordering=True,
+            )
+            result = im.call(self.visit(node.func, **kwargs))(
+                *lowered_args, *lowered_kwargs.values()
+            )
+
+            # scan operators return an iterator of tuples, transform into tuples of iterator again
+            if isinstance(node.func.type, ts_ffront.ScanOperatorType):
+                raise NotImplementedError("TODO")
+
+            return result
+
+        raise AssertionError(
+            f"Call to object of type '{type(node.func.type).__name__}' not understood."
+        )
+
+    def _visit_astype(self, node: foast.Call, **kwargs: Any) -> itir.Expr:
+        assert len(node.args) == 2 and isinstance(node.args[1], foast.Name)
+        obj, new_type = self.visit(node.args[0], **kwargs), node.args[1].id
+
+        def create_cast(expr: itir.Expr) -> itir.FunCall:
+            return im.as_fieldop(
+                im.lambda_("__val")(im.call("cast_")(im.deref("__val"), str(new_type)))
+            )(expr)
+
+        if not isinstance(node.type, ts.TupleType):
+            return create_cast(obj)
+
+        return lowering_utils.process_elements(create_cast, obj, node.type)
+
+    def _visit_where(self, node: foast.Call, **kwargs: Any) -> itir.FunCall:
+        if not isinstance(node.type, ts.TupleType):
+            return im.op_as_fieldop("if_")(*self.visit(node.args))
+
+        cond_ = self.visit(node.args[0])
+        cond_symref_name = f"__cond_{eve_utils.content_hash(cond_)}"
+
+        def create_if(cond: itir.Expr) -> Callable[[itir.Expr, itir.Expr], itir.FunCall]:
+            def create_if_impl(true_: itir.Expr, false_: itir.Expr) -> itir.FunCall:
+                return im.op_as_fieldop("if_")(cond, true_, false_)
+
+            return create_if_impl
```

Comment:
```suggestion
          def create_if(true_: itir.Expr, false_: itir.Expr) -> itir.FunCall:
              return im.op_as_fieldop("if_")(im.ref(cond_symref_name), true_, false_)
```
I consider closures a feature^^

Reply:
done